### PR TITLE
Enable JSON parsing of variables that are JSON strings

### PR DESCRIPTION
**Pull request by benjdlambert**
## Hey, I just made a Pull Request!

enable the JSON parsing of the response from templated variables in the `v2beta1` syntax. Previously if template parameters json strings they were left as strings, they are now parsed as JSON objects.

Before:

```yaml
- id: publish
  name: Publish
  action: publish:github
  input:
    destination: '{"host":"github.com","owner":"owner","repo":"repo"}'
```

Now:

```yaml
- id: publish
  name: Publish
  action: publish:github
  input:
    destination:
      host: 'github.com'
      owner: 'owner'
      repo: 'repo'
```

Deprecates the `repoUrl` parameters and has been replaced in favour of `destination` which means that you'll need to update your templates where you use the `RepoUrlPicker`. Firstly, updating the `parameters` section:

```diff
  parameters:
    - title: Choose a location
      required:
-        - repoUrl
+        - destination
      properties:
-        repoUrl:
+        destination:
          title: Repository Location
          type: string
          ui:field: RepoUrlPicker
          ui:options:
            allowedHosts:
              - github.com
```

And then updating the steps as follows:

```diff
  steps
    - id: publish
      name: Publish
      action: publish:github
      input:
        allowedHosts: ['github.com']
        description: 'This is {{ parameters.name }}'
-        repoUrl: '{{ parameters.repoUrl }}'
+        destination: '{{ parameters.destination }}'
```

This will need to be done for **each** publish action in your templates!

#### :heavy_check_mark: Checklist

 

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ import {
   ScmIntegrationRegistry,
 } from '@backstage/integration';
 import { initRepoAndPush } from '../../../stages/publish/helpers';
-import { getRepoSourceDirectory, parseRepoUrl } from './util';
+import { getRepoSourceDirectory, Destination } from './util';
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts:23]: possibly rename `Destination` to something like `ScmRepository` or so?

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts [hunk 2]
```diff
@@ -37,20 +37,37 @@ export function createPublishGitlabAction(options: {
...
           repoVisibility: {
-            title: 'Repository Visiblity',
+            title: 'Repository Visibility',
             type: 'string',
             enum: ['private', 'public', 'internal'],
           },
           sourcePath: {
             title:
-              'Path within the workspace that will be used as the repository root. If omitted, the entire workspace will be published as the respository.',
+              'Path within the workspace that will be used as the repository root. If omitted, the entire workspace will be published as the Repository.',
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:70]:
```suggestion
              'Path within the workspace that will be used as the repository root. If omitted, the entire workspace will be published as the repository.',
```

**Review by freben — APPROVED** [review]
some nits but then approv'd

**Review by Rugvip — COMMENTED** [review]
This doesn't deprecate `repoUrl` as much as it completely removes them 😅. Thoughts on trying to create a bit of backwards compatibility so that we don't break all templates out there when people upgrade?

👍 to overall design. I'm thinking we can also work this into a solution to https://github.com/backstage/backstage/discussions/5028. Right now you kinda rely on JSON being passed from the parameters, which ofc is a bit shite for humans to write. We could add a handlebars helper that lets you serialize a set of input parameters into JSON, just `handlebars.registerHelper('json', data => JSON.stringify(data)`. That way you can take a whole group of input parameters and pass them on as a single input to an action.

And let's add support for `[...]` in addition to `{...}` right? 😁

**Comment by benjdlambert** [comment]
Hmm yeah - wondering how hard that's going to be though. It's going to complicate the publishers a fair amount in order to support both but maybe we should do. With the stability index of 1 still for the scaffolder thought that it would have been fine to go this way but happy to move this logic to another component entirely and make repoUrl optional and also destination optional in the publishers if that's what you mean?

**Pull request closed without merging**